src/core/Tree/TreeNode.tsx

```diff
@@ -0,0 +1,158 @@
+  const {
+    label,
+    sublabel,
+    children,
+    style,
+    className,
+    icon,
+    isDisabled = false,
+    ...rest
+  } = props;
+  useTheme();
+
+  const context = React.useContext(TreeContext);
+  const currentDepth = context?.nodeDepth ?? 0;
+
+  const styleLevel = React.useMemo(
+    () =>
+      getWindow()?.CSS?.supports?.(`--level: ${currentDepth}`)
+        ? { '--level': currentDepth, ...style }
+        : { marginLeft: currentDepth ? currentDepth * 28 : 0, ...style },
+    [currentDepth, style],
+  );
+
+  const [expanded, setExpanded] = React.useState(false);
+  const [isChecked, setIsChecked] = React.useState(
+    React.isValidElement(context?.checkbox)
+      ? context?.checkbox.props['checked']
+      : false,
+  );
+  const isSelected = React.useMemo(() => {
+    return context?.selectedNode === label;
+  }, [context?.selectedNode, label]);
+
+  return (
+    <li role='treeitem' aria-expanded={expanded} {...rest}>
+      <div
+        className={cx('iui-tree-node', {
+          'iui-active': isSelected,
+          'iui-disabled': isDisabled,
+          className,
+        })}
+        style={styleLevel}
+        onClick={(e) => {
+          if (e.currentTarget != e.target) {
+            return;
+          }
+          if (isSelected) {
+            context?.setSelectedNode?.('');
+          } else if (!isDisabled) {
+            context?.setSelectedNode?.(label);
+            context?.onNodeSelected?.();
+          }
+        }}
+      >
+        {context?.checkbox && React.isValidElement(context?.checkbox)
+          ? React.cloneElement(context?.checkbox, {
+              className: 'iui-tree-node-checkbox',
+              disabled: isDisabled,
+              checked: isChecked,
+              onClick: () => {
+                setIsChecked(!isChecked);
+                context.onNodeCheckboxSelected?.();
+              },
+            })
+          : context?.checkbox}
+        <div className='iui-tree-node-content'>
+          {children && (
+            <IconButton
+              styleType='borderless'
+              size='small'
+              onClick={() => {
+                setExpanded(!expanded);
+                context?.onNodeExpanded?.();
+              }}
+              disabled={isDisabled}
+            >
+              <SvgChevronRight
+                className={cx('iui-tree-node-content-expander-icon', {
+                  'iui-tree-node-content-expander-icon-expanded': expanded,
+                })}
+              />
+            </IconButton>
+          )}
+          {icon &&
+            React.cloneElement(icon, {
+              className: cx('iui-tree-node-content-icon', icon.props.className),
+            })}
+          <span className='iui-tree-node-content-label'>
+            <div className='iui-tree-node-content-title'>{label}</div>
+            <div className='iui-tree-node-content-caption'>{sublabel}</div>
+          </span>
+        </div>
+      </div>
+      {children && expanded && (
```

Comment:
Would you recommend having a flat list? Because we were fallowing W3 guidelines for having nested structure with all required roles. Also, Material UI is having the same. https://www.w3.org/TR/wai-aria-practices-1.2/#TreeView

Reply:
A flat list is much easier to virtualize, but list virtualization always comes at the cost of accessibility. I cannot find any resources that talk about virtualized list accessibility, and I expect all accessibility tools to have poor support for this pattern. I raised this point because I saw virtualization is being implemented for the table component.

If iTwinUI team chose not to support virtualization in any of its components, I would consider it as a big win for all end users, because scrolling an enormous list with a tiny scrollbar is always a horrible experience. Absence of out-of-the-box virtualization may lead developers to rethink data presentation strategy and come up with something that has friendlier UX.

Reply:
@roluk  that is a really good comment, thank you! but.. We already have virtualized trees used in our itwin.js applications to show those complicated data structures and I do not think they would replace that UI into something else. But we also want our tree to be used there as we take care of visual issues which are present currently and ensure quality.
We will have virtualization in at least table, I think, cause this is waited and required. Even though after your comment, we might rethink something and try to see if people are up to using pagination :D
So not sure, what would be ok way to go here :/